src/scripts/list-groups.ts

```diff
@@ -1,82 +1,213 @@
-  return table(
+const formatGroupsTable: (groups: readonly Group[]) => string = (
+  groups
+): string =>
+  table(
     [
       ['ID', 'Name', 'Type'],
-      ...groups.map((group: Group) => [group.id, group.name, group.type]),
+      ...groups.map(({ id, name, type }: Group) => [id, name, type]),
     ],
     {
-      // eslint-disable-next-line functional/functional-parameters
-      drawHorizontalLine: () => false,
-      // eslint-disable-next-line functional/functional-parameters
-      drawVerticalLine: () => false,
+      drawHorizontalLine: constant(false),
+      drawVerticalLine: constant(false),
     }
   );
+
+/**
+ * Okta API scopes relevant to the `list-groups` command. See Okta Developer
+ * Documentation for {@link https://developer.okta.com/docs/api/openapi/okta-management/management/tag/Group/ Groups}
+ * and {@link https://developer.okta.com/docs/api/openapi/okta-management/management/tag/User/ Users}.
+ */
+type Scope = 'groups' | 'users';
+
+/**
+ * User ID option.
+ *
+ * @example
+ * ```typescript
+ * const userIdOption: UserIdOption = O.some('00gddktac01w2dgmL5d7');
+ * ```
+ */
+export type UserIdOption = O.Option<string>;
+
+/**
+ * Parsed `list-groups` command options.
+ */
+type ParsedListGroupsOptions = {
+  readonly oktaConfiguration: OktaConfiguration;
+  readonly scopes: readonly Scope[];
+  readonly userIdOption: UserIdOption;
+};
+
+/**
+ * Parses the `list-groups` command options. If the user ID option is not
+ * provided, then the `groups` scope is sufficient. Otherwise, the `users` scope
+ * is required.
+ *
+ * @param listGroupsOptions The client ID, private key, organisation URL, and
+ * user ID options passed to the `list-groups` command.
+ * @returns The parsed `list-groups` options, including the Okta configuration,
+ * Okta API scopes, and user ID option.
+ */
+const parseListGroupsOptions: (
+  listGroupsOptions: ListGroupsOptions
+) => ParsedListGroupsOptions = ({ clientId, privateKey, orgUrl, userId }) =>
+  pipe(userId, O.fromNullable, (userIdOption) => ({
+    oktaConfiguration: {
+      clientId,
+      privateKey,
+      orgUrl,
+    },
+    scopes: O.match(
+      constant<readonly Scope[]>(['groups']),
+      constant<readonly Scope[]>(['users'])
+    )(userIdOption),
+    userIdOption,
+  }));
+
+/**
+ * Okta client instance and user ID option.
+ */
+type OktaClientAndUserIdOption = {
+  readonly client: Client;
+  readonly userIdOption: UserIdOption;
+};
+
+/**
+ * Creates an Okta client instance and user ID option.
+ *
+ * @param parsedListGroupsOptions The parsed `list-groups` options, including
+ * the Okta configuration, Okta API scopes, and user ID option.
+ *
+ * @returns The Okta client and user ID option.
+ */
+const createOktaClientAndUserIdOption: (
+  parsedListGroupsOptions: ParsedListGroupsOptions
+) => OktaClientAndUserIdOption = ({
+  oktaConfiguration,
+  scopes,
+  userIdOption,
+}) => ({
+  client: oktaReadOnlyClient(oktaConfiguration, scopes),
+  userIdOption,
```

Comment:
**Nit:** having to return a parameter like this is a bit of a code smell. The caller has this information already, it normally shouldn't be necessary to plumb it through like this.

Reply:
I think I was trying to make it so I didn't have to declare a variable and could have a clean pipeline of functions for aesthetic reasons. I couldn't do it without these functions because `userId` (or `userIdOption`, as it later became) was required in two places. But I can see now it just makes things more complicated. I'll refactor the code.

Reply:
These steps aren't benefiting from the functional pipeline that I can see (i.e. they aren't dealing in `TaskEither`). If so, better to rewrite to be imperative.

Where they do benefit from being in the pipeline, I reckon using [bind](https://gcanti.github.io/fp-ts/modules/TaskEither.ts.html#bind) | [let](https://gcanti.github.io/fp-ts/modules/TaskEither.ts.html#let) is a better approach than this.

Reply:
Right I see. I haven't got much experience using `bind` or `let`, but will keep them in mind.